src/mito2/src/memtable/merge_tree/merger.rs

```diff
@@ -0,0 +1,633 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::{Ordering, Reverse};
+use std::collections::BinaryHeap;
+use std::fmt::Debug;
+use std::ops::Range;
+
+use datatypes::arrow::array::{
+    ArrayRef, TimestampMicrosecondArray, TimestampMillisecondArray, TimestampNanosecondArray,
+    TimestampSecondArray, UInt64Array,
+};
+use datatypes::arrow::datatypes::{DataType, TimeUnit};
+
+use crate::error::Result;
+use crate::memtable::merge_tree::data::{DataBatch, DataBufferReader, DataPartReader};
+use crate::memtable::merge_tree::PkIndex;
+
+pub trait Item: Clone {
+    type Key: Ord;
+
+    /// Remaining rows in item.
+    fn remaining(&self) -> usize;
+
+    /// Whether current item is exhausted.
+    fn is_empty(&self) -> bool;
+
+    /// The key range of item.
+    fn current_range(&self) -> Range<Self::Key>;
+
+    /// Searches given key in current item.
+    fn search_key(&self, key: &Self::Key) -> std::result::Result<usize, usize>;
+
+    /// Slice item.
+    fn slice(&self, range: Range<usize>) -> Self;
+}
+
+pub trait Node: Ord {
+    type Item: Item;
+
+    /// Returns current item of node and fetch next.
+    fn fetch_next(&mut self) -> Result<Self::Item>;
+
+    /// Returns true if current node is not exhausted.
+    fn is_valid(&self) -> bool;
+
+    /// Current item of node.
+    fn current_item(&self) -> &Self::Item;
+
+    /// Whether the other node is behind (exclusive) current node.
+    fn is_behind(&self, other: &Self) -> bool;
+
+    /// Skips first `num_to_skip` rows from node's current batch. If current batch is empty it fetches
+    /// next batch from the node.
+    ///
+    /// # Panics
+    /// If the node is EOF.
+    fn skip(&mut self, offset_to_skip: usize) -> Result<()>;
+}
+
+pub struct Merger<T, I> {
+    heap: BinaryHeap<T>,
+    current_item: Option<I>,
+}
+
+impl<T, I> Merger<T, I>
+where
+    I: Item + Debug,
+    T: Node<Item = I>,
+{
+    pub(crate) fn try_new(nodes: Vec<T>) -> Result<Self> {
+        let mut heap = BinaryHeap::with_capacity(nodes.len());
+        for node in nodes {
+            if node.is_valid() {
+                heap.push(node);
+            }
+        }
+        let mut merger = Merger {
+            heap,
+            current_item: None,
+        };
+        merger.next()?;
+        Ok(merger)
+    }
+
+    /// Returns true if current merger is still valid.
+    pub(crate) fn is_valid(&self) -> bool {
+        self.current_item.is_some()
+    }
+
+    /// Advances current merger to next item.
+    pub(crate) fn next(&mut self) -> Result<()> {
+        if let Some(mut top_node) = self.heap.pop() {
```

Comment:
nitpick: Consider using `let-else`.

Reply:
> nitpick: Consider using `let-else`.

fixed in https://github.com/GreptimeTeam/greptimedb/pull/3361/commits/dc1c86a1ff2acf1c0cea69dc3960a661de2f7a23#diff-8e513b995e680c513011a438989b7551954f873b55a1a4ba602550e21d5d8094R101-R105